Docs feature branch: SQLite in DO (#16903)
cloudflare/cloudflare-docs: src/content/partials/durable-objects/durable-objects-vs-d1.mdx

---
{}
---

> ## SQL in Durable Objects vs D1

Author: cc @joshthoward @justin-mp

Cloudflare Workers offers a SQLite-backed serverless database product - [D1](/d1/). How should you compare [SQLite in Durable Objects](/durable-objects/best-practices/access-durable-objects-storage/#sql-storage) and D1?

> **D1 should be viewed as a managed database product.** 

Reviewer: ```suggestion
**D1 is a managed database product.** 
```

> D1 fits into a familiar architecture for developers, where application servers communicate with a database over the network. Application servers are typically Workers; however, D1 also supports external, non-Worker access via an [HTTP API](https://developers.cloudflare.com/api/operations/cloudflare-d1-query-database). 

Reviewer 2: nit: I think it's a missed opportunity not to highlight the framework support that D1 has and that DO probably cannot have.

> D1 fits into a familiar architecture for developers, where application servers communicate with a database over the network. Application servers are typically Workers; however, D1 also supports external, non-Worker access via [HTTP API](https://developers.cloudflare.com/api/operations/cloudflare-d1-create-database). 

Reviewer: Given that it doesn't seem possible to point to all of the D1 HTTP API methods as a group, it's probably better to point to querying the database, as that's actually a data plane operation.

```suggestion
D1 fits into a familiar architecture for developers, where application servers communicate with a database over the network. Application servers are typically Workers; however, D1 also supports external, non-Worker access via an [HTTP API](https://developers.cloudflare.com/api/operations/cloudflare-d1-query-database). 
```

> D1 aims for a "batteries included" feature set, including HTTP API, [database schema management](/d1/reference/migrations/#_top), [data import/export](/d1/build-with-d1/import-export-data/), database query [insights](/d1/observability/metrics-analytics/#query-insights), and upcoming built-in read replication. 

Reviewer: We shouldn't mention unreleased features in documentation.

```suggestion
D1 aims for a "batteries included" feature set, including the above HTTP API, [database schema management](/d1/reference/migrations/#_top), [data import/export](/d1/build-with-d1/import-export-data/), database query and [insights](/d1/observability/metrics-analytics/#query-insights). 
```

> With D1, your application code and SQL database queries are not colocated which can impact application performance. Workers has [Smart Placement](/workers/configuration/smart-placement/#_top) to dynamically run your Worker in the best location to reduce total Worker request latency, considering everything your Worker talks to, including D1.

Reviewer: ```suggestion
With D1, your application code and SQL database queries are not colocated which can impact application performance. If performance is a concern with D1, Workers has [Smart Placement](/workers/configuration/smart-placement/#_top) to dynamically run your Worker in the best location to reduce total Worker request latency, considering everything your Worker talks to, including D1.
```

> **SQLite in Durable Objects is a lower-level compute with storage building block for distributed systems.** By design, Durable Objects are accessed with Workers-only.

Reviewer 3: ```suggestion
**SQLite in Durable Objects is a lower-level compute with storage building block for distributed systems.**

By design, Durable Objects are accessed with Workers-only.
```

> Durable Objects require a bit more effort, but in return, give you more flexibility and control. With Durable Objects, you have two pieces of code that run in different places: a front-end Worker which routes incoming requests from the Internet to a unique Durable Object, and the Durable Object itself, which runs on the same machine as the SQLite database. 

Reviewer: ```suggestion
Durable Objects require a bit more effort, but in return, give you more flexibility and control. With Durable Objects, you must implement two pieces of code that run in different places: a front-end Worker which routes incoming requests from the Internet to a unique Durable Object, and the Durable Object itself, which runs on the same machine as the SQLite database.  You get to choose what runs where and it may be that your application benefits from running some application business logic right next to the database.
```

> You may need to carefully consider where to run different application code, and you may need to build some of your own database tooling that exists out-of-the-box with D1. But because you are in full control, you can tailor the solution to your application's needs and potentially achieve more.

Reviewer: ```suggestion
With SQLite in Durable Objects, you may also need to build some of your own database tooling that comes out-of-the-box with D1.
```
